lib/reconstruct.ts

```diff
@@ -0,0 +1,73 @@
+import { types, Root, Token, Tokens, Group } from './types';
+import * as sets from './sets';
+
+const simplifications: [sets.SetFunc, string][] = [
+  [sets.words, '\\w'],
+  [sets.notWords, '\\W'],
+  [sets.ints, '\\d'],
+  [sets.notInts, '\\D'],
+  [sets.whitespace, '\\s'],
+  [sets.notWhitespace, '\\S'],
+  [sets.anyChar, '.']
+];
+
+const reduceStack = (stack: Token[]): string => stack.map(reconstruct).join('');
+
+const createAlternate = (token: Root | Group): string => {
+  if ('options' in token) {
+    return token.options.map(reduceStack).join('|');
+  } else if ('stack' in token) {
+    return reduceStack(token.stack);
+  } else {
+    throw new Error(`options or stack must be Root or Group token`);
+  }
+};
+
+export const reconstruct = (token: Tokens): string => {
+  switch (token.type) {
+    case types.ROOT:
+      return createAlternate(token);
+    case types.CHAR:
+      const c = String.fromCharCode(token.value);
+      return (/[[\]^.\/|?*+()]/.test(c) ? '\\' : '') + c;
```

Comment:
Missing curly braces, hyphen, backslash, `^`, `$`?

Reply:
Presumably it should be the same as this https://github.com/fent/ret.js/blob/47e88c3194641c39b5a82c60a7c6e1fa6a813af1/lib/util.ts#L37, however I note that there backslash and hyphen are not escaped, if they should be I can refactor this all into the function
```ts
function escapeSpecial(c: string) {
  return /[[\]{}^$.|?*+()]\\\-/.test(c) ? '\\' + c : c;
}
```
and use this in both the util and reconstruct file

Reply:
I’m not sure what `strToChars` is meant for. `lbs` looks suspicious (only one level of escaping checked?). It matches case-insensitive regex components case-sensitively. It matches octal escapes that are two digits long, but they can be three digits. `\c@`, `\c[`, `\c]`, `\c\`, `\c?` don’t exist, but `[\c<digit>]` does, `\b` doesn’t have to appear alone in a character class…